### PR TITLE
Bugs fix and refacto

### PR DESCRIPTION
**Pull request by Gauddel**

### CODE REVIEW

**Comment by hilmarx** [comment]
for `_isVaultUnsafeOSM`, instadapp will input 0.67*10**18 for the `_minColRatio` right?

File: contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol [hunk 1]
```diff
@@ -2,41 +2,41 @@
...
-    }
+    uint256 _colAmtInETH;
 
-    uint256 colPrice = abi.decode(returndata, (uint256));
+    AaveUserData memory userData = _getUserData(_dsa);
 
-    AaveUserData memory userData = _getPosition(_dsa);
-    AaveTokenData memory tokenData = _collateralData(_debtToken);
+    if (_priceChainlinkOracle == ETH_ETH_PRICEFEEDER) _colAmtInETH = _colAmt;
+    else _colAmtInETH = wmul(_colAmt, _getPrice(_priceChainlinkOracle));
```
Comment (hilmarx) [inline contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol:30]: @Gauddel please explain this. Did you include that in order to enable other collateral types than ETH?

If so, why? Because Aave always returns every collateral and debt value denominated in ETH?

Comment (hilmarx) [reply contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol:30]: The result here will be that Instadapp has to always make sure they pass the correct Chainlink oracle address via the payload for e.g. `BAT` right? Obviously then the Oracle Aggregator would have been easier for them.

File: contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol [hunk 2]
```diff
@@ -2,41 +2,41 @@
...
-        ) <= tokenData.ltv;
+        wdiv(
+            userData.totalCollateralETH +
+                _percentMul(
+                    _colAmtInETH,
+                    _getAssetLiquidationThreshold(_colToken)
+                ),
+            userData.totalBorrowsETH +
+                wmul(_debtAmt, _getPrice(DAI_ETH_PRICEFEEDER))
+        ) > 1e18;
```
Comment (hilmarx) [inline contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol:41]: @Gauddel I dont get this, need to explain in person. We dont take into account gelato and instadapp fees here, do we?

**Comment by hilmarx** [comment]
We stil dont take into account gelato and instadapp fees in `ConditionCanDoRefinance` correct? I thought we discussed this.

File: contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol (hunk 1 above)
Comment (Gauddel) [reply contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol:30]: Yes aave return user data in ether, but the col we want to deposit in aave protocol can be something else than ether right?

File: contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol (hunk 2 above)
Comment (Gauddel) [reply contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol:41]: yes we do, debAmt take into account gelato and instadapp fees,  check FCanDoRefinance.

File: contracts/contracts/resolvers/MakerResolver.sol [hunk 3]
```diff
@@ -67,7 +67,8 @@ contract MakerResolver {
         IVat vat = IVat(manager.vat());
         bytes32 ilk = _stringToBytes32(_colType);
         (uint256 art, uint256 rate, , uint256 line, ) = vat.ilks(ilk);
-        return line - (art * rate);
+
+        return (line / rate) - art;
```
Comment (hilmarx) [inline contracts/contracts/resolvers/MakerResolver.sol:71]: can you explain this change here?

File: contracts/functions/dapps/FAave.sol [hunk 4]
```diff
@@ -48,44 +49,40 @@ function _getUserData(
...
+        .getReserveConfigurationData(_getTokenAddr(_token));
 }
 
-function _percentDiv(uint256 value, uint256 percentage) pure returns (uint256) {
-    require(percentage != 0, "Division Error");
-    uint256 percentageFactor = 1e4; //percentage plus two decimals
-    uint256 halfPercentage = percentage / 2;
+// Aave specific math
+function _percentMul(uint256 _value, uint256 _percentage)
+    pure
```
Comment (hilmarx) [inline contracts/functions/dapps/FAave.sol:68]: can you explain this change please?

Comment (hilmarx) [inline contracts/functions/dapps/FAave.sol:68]: can you explain this change please?

File: contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol (hunk 1 above)
Comment (hilmarx) [reply contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol:30]: yes exactly, then it makes sense

Comment (hilmarx) [reply contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol:30]: Would be significantly easier for instadapp to manage if we would use the OracleAggregator, as now for ETH they will have to pass `address(0)` and for other collateral types the respective oracle addresses, which is the whole point why we wrote the oracle aggregator. 

We can leave it like that, but this will add complexity and therefore potential room for errors on Instadapps side, because if they mess up the inputs the transaction will likely revert. 

I would use the OracleAggregator. @gitpusha @Gauddel thoughts?

File: contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol (hunk 2 above)
Comment (hilmarx) [reply contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol:41]: can you please explain the maths here for dummies? Why do we _percentMul colAmtInEth with the liquidation threshold?

Comment (Gauddel) [reply contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol:41]: I saw a simplification, i will push something.

File: contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol (hunk 1 above)
Comment (gitpusha) [reply contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol:30]: I would think that it should be easy to implement a small in-memory oracle mapping off-chain in InstaDapps UI for this kind of stuff. 

But if the OracleAggregator is trustless and safe to use, then it makes sense here. 

But if its trustless (immutable) then it cannot be upgraded and hence if an oracle change is needed, we would have to redeploy everything. 

I think it's a hard call. We should probably ask InstaDapp what they prefer and let them make this decision. 

I guess an upgradeable oracle aggregator managed by us is convenient here but trusted. Trusted that we don't make mistakes or grief willingly.

File: contracts/functions/dapps/FAave.sol (hunk 4 above)
Comment (gitpusha) [reply contracts/functions/dapps/FAave.sol:68]: duplicate

File: contracts/contracts/connectors/debt_bridge/from_maker/ConnectGelatoDataMakerToCompound.sol [hunk 5]
```diff
@@ -169,20 +166,17 @@ contract ConnectGelatoDataMakerToCompound is
...
         (address[] memory _targets, bytes[] memory _datas) =
             _spellsMakerToCompound(
                 _vaultId,
                 _colToken,
                 daiToBorrow,
                 _getMakerVaultCollateralBalance(_vaultId),
-                gasFeesPaidFromBor
+                wdiv(
+                    _getGelatoExecutorFees(_getGasCostMakerToCompound(route)),
+                    _getPrice(DAI_ETH_PRICEFEEDER)
```
Comment (gitpusha) [inline contracts/contracts/connectors/debt_bridge/from_maker/ConnectGelatoDataMakerToCompound.sol:178]: probably should rename `_getPrice` to `_getChainlinkPrice` for more readability

File: contracts/contracts/connectors/debt_bridge/from_maker/ConnectGelatoDataMakerToMaker.sol [hunk 6]
```diff
@@ -127,11 +124,10 @@ contract ConnectGelatoDataMakerToMaker is
     function getDataAndCastMakerToMaker(
         uint256 _vaultAId,
         uint256 _vaultBId,
-        string calldata _colType,
-        address _colToken
+        string calldata _colType
```
Comment (gitpusha) [inline contracts/contracts/connectors/debt_bridge/from_maker/ConnectGelatoDataMakerToMaker.sol:127]: technically in order to ensure that there is no revert here, we would have to make sure that `_colType` is a correct one. Otherwise `_encodeOpenMakerVault` will revert. 

but I think we can take this risk for sake of not adding more stuff to the equation.

File: test/integration/debt_bridge/from_maker/functionality/1_ETHAToNewETHB.test.js [hunk 7]
```diff
@@ -219,13 +219,15 @@ describe("Full Debt Bridge refinancing loan from ETH-A => new ETH-B", function (
...
-      )
-    )[0];
+    const daiETHPrice = await contracts.chainlinkResolver.getPrice(
+      constants.DAI_ETH_PRICEFEEDER
+    );
+
+    const gasFeesPaidFromDebt = ethers.BigNumber.from(gasCost)
+      .mul(gelatoGasPrice)
+      .mul(ethers.utils.parseUnits("1", 18))
+      .add(daiETHPrice.div(ethers.utils.parseUnits("2", 0)))
```
Comment (gitpusha) [inline test/integration/debt_bridge/from_maker/functionality/1_ETHAToNewETHB.test.js:229]: why do we have this?

File: test/integration/debt_bridge/from_maker/helpers/constants.js [hunk 8]
```diff
@@ -2,7 +2,7 @@ const hre = require("hardhat");
 const { ethers } = hre;
 
 const ETH = "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE";
-const GAS_LIMIT = "5000000";
+const GAS_LIMIT = "6000000";
```
Comment (gitpusha) [inline test/integration/debt_bridge/from_maker/helpers/constants.js:5]: why was this change needed? did you go beyond 5 mio gas?

File: contracts/vendor/Convert.sol [hunk 9]
```diff
@@ -13,3 +13,10 @@ function _stringToBytes32(string memory str) pure returns (bytes32 result) {
 function _convertTo18(uint256 _dec, uint256 _amt) pure returns (uint256 amt) {
     amt = _mul(_amt, 10**(18 - _dec));
 }
+
+function _convertIntToUint(int256 a) pure returns (uint256) {
+    if (a >= 0) {
+        return uint256(a);
+    }
+    return uint256(a * -1);
```
Comment (gitpusha) [inline contracts/vendor/Convert.sol:21]: Should we not rather `revert` in `else` statement as we are only expecting `uints` here really?

File: contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol (hunk 1 above)
Comment (hilmarx) [reply contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol:30]: The Oracle Contract is already owned by the multisig, so making mistakes here will be hard.

File: contracts/contracts/connectors/debt_bridge/from_maker/ConnectGelatoDataMakerToMaker.sol (hunk 6 above)
Comment (hilmarx) [reply contracts/contracts/connectors/debt_bridge/from_maker/ConnectGelatoDataMakerToMaker.sol:127]: Yes I think that is a griefing vector, but we should not optmize for it now, will be handled in the new version by just checking off chain before the execution.

File: contracts/contracts/resolvers/MakerResolver.sol (hunk 3 above)
Comment (Gauddel) [reply contracts/contracts/resolvers/MakerResolver.sol:71]: This function is used for knowing how much debt we can still borrow without the stability fees, so the calculation should not take into account this fees which is represented by rate

File: contracts/functions/dapps/FAave.sol (hunk 4 above)
Comment (Gauddel) [reply contracts/functions/dapps/FAave.sol:68]: Resolved, we didn't use it anymore

File: contracts/contracts/connectors/debt_bridge/from_maker/ConnectGelatoDataMakerToCompound.sol (hunk 5 above)
Comment (Gauddel) [reply contracts/contracts/connectors/debt_bridge/from_maker/ConnectGelatoDataMakerToCompound.sol:178]: yes we can rename it like that

File: contracts/contracts/connectors/debt_bridge/from_maker/ConnectGelatoDataMakerToMaker.sol (hunk 6 above)
Comment (Gauddel) [reply contracts/contracts/connectors/debt_bridge/from_maker/ConnectGelatoDataMakerToMaker.sol:127]: Resolved?

File: test/integration/debt_bridge/from_maker/helpers/constants.js (hunk 8 above)
Comment (Gauddel) [reply test/integration/debt_bridge/from_maker/helpers/constants.js:5]: Yes sometimes when we do aave (flashloan) + aave (refinance) or maker (flashloan) + aave (refinance)

File: contracts/vendor/Convert.sol (hunk 9 above)
Comment (Gauddel) [reply contracts/vendor/Convert.sol:21]: We can revert, even if this use case should normally not happen. It's better to revert here i think.

File: test/integration/debt_bridge/from_maker/functionality/1_ETHAToNewETHB.test.js (hunk 7 above)
Comment (Gauddel) [reply test/integration/debt_bridge/from_maker/functionality/1_ETHAToNewETHB.test.js:229]: the add is for precisio, sometimes the test fail due to this lack of precision. We are doing the same thing in DSMath.

File: contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol (hunk 2 above)
Comment (hilmarx) [reply contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol]: Resolved.

File: contracts/functions/dapps/FAave.sol (hunk 4 above)
Comment (hilmarx) [reply contracts/functions/dapps/FAave.sol]: Resolved.

File: contracts/contracts/connectors/debt_bridge/from_maker/ConnectGelatoDataMakerToMaker.sol (hunk 6 above)
Comment (hilmarx) [reply contracts/contracts/connectors/debt_bridge/from_maker/ConnectGelatoDataMakerToMaker.sol:127]: Resolved.

Comment (gitpusha) [reply contracts/contracts/connectors/debt_bridge/from_maker/ConnectGelatoDataMakerToMaker.sol:127]: yes - although `ConditionMakerToMakerSafe.destVaultWillBeSafe` technically is not a secure sanity check as the `_colType` could be a correct one there but an incorrect one in the actionPayload. 

but I think we can leave this one to be a griefing edge case.

File: contracts/contracts/resolvers/MakerResolver.sol (hunk 3 above)
Comment (hilmarx) [reply contracts/contracts/resolvers/MakerResolver.sol:71]: This change is not affecting anything on-chain. Resolved.

Comment (hilmarx) [reply contracts/contracts/resolvers/MakerResolver.sol:71]: This change is not affecting anything on-chain. Resolved.

File: contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol (hunk 1 above)
Comment (gitpusha) [reply contracts/functions/gelato/conditions/aave/FAavePositionWillBeSafe.sol:29]: Resolved in favor of OracleAggregator
